### PR TITLE
Update to use the new parent information in fparser and remove white space after generated "END DO/IF" (for 634)

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
This is pretty much ready to go but is waiting on the finalisation of https://github.com/stfc/fparser/pull/230

**Comment by arporter** [comment]
Ready for first review from @rupertford. Note that although the diff looks big, a lot of it is just due to the removal of white space after 'END DO' and 'END IF'.

**Comment by arporter** [comment]
Done. Some of the new calls to `process_nodes` that were brought in by the merge were still passing `nodes_parent` and that doesn't exist anymore.

File: src/psyclone/tests/nemo/transformations/profiling/nemo_profile_test.py [hunk 1]
```diff
@@ -393,7 +393,8 @@ def test_profiling_missing_end(parser):
     loops[0]._ast_end = psy._ast
     with pytest.raises(InternalError) as err:
         _ = psy.gen
-    assert "Failed to find the location of 'PROGRAM do_loop" in str(err.value)
+    assert ("nodes of the profiling region in the fparser2 parse tree do not "
```
Comment (rupertford) [inline src/psyclone/tests/nemo/transformations/profiling/nemo_profile_test.py]: 2020 copyright date please.

File: src/psyclone/tests/psyir/frontend/fparser2_test.py [hunk 2]
```diff
@@ -880,14 +880,14 @@ def test_parse_array_dimensions_unhandled(monkeypatch):
 
     def walk_ast_return(_1, _2, _3=None, _4=None):
         '''Function that returns a unique object that will not be part
-        of the implemented handling in the walk_ast method caller.'''
+        of the implemented handling in the walk method caller.'''
```
Comment (rupertford) [inline src/psyclone/tests/psyir/frontend/fparser2_test.py]: 2020 copyright date please.

File: src/psyclone/tests/psyir/frontend/literals_test.py [hunk 3]
```diff
@@ -60,7 +60,7 @@ def test_handling_literal(code, dtype):
     astmt = Fortran2003.Assignment_Stmt(reader)
     fake_parent = Node()
     processor = Fparser2Reader()
-    processor.process_nodes(fake_parent, [astmt], None)
+    processor.process_nodes(fake_parent, [astmt])
```
Comment (rupertford) [inline src/psyclone/tests/psyir/frontend/literals_test.py]: 2020 ...

File: src/psyclone/tests/psyir/frontend/size_intrinsic_test.py [hunk 4]
```diff
@@ -55,7 +55,7 @@ def test_size(expression, parser):
     processor = Fparser2Reader()
     reader = FortranStringReader(expression)
     fp2intrinsic = Execution_Part(reader).content[0]
-    processor.process_nodes(fake_parent, [fp2intrinsic], None)
+    processor.process_nodes(fake_parent, [fp2intrinsic])
```
Comment (rupertford) [inline src/psyclone/tests/psyir/frontend/size_intrinsic_test.py]: 2020 ...

File: src/psyclone/tests/psyir/frontend/where_handler_test.py [hunk 5]
```diff
@@ -65,9 +65,9 @@ def process_where(code, fparser_cls):
     reader = FortranStringReader(code)
     fparser2spec = fparser_cls(reader)
     if fparser_cls is Fortran2003.Execution_Part:
-        processor.process_nodes(sched, fparser2spec.content, None)
+        processor.process_nodes(sched, fparser2spec.content)
```
Comment (rupertford) [inline src/psyclone/tests/psyir/frontend/where_handler_test.py]: 2020 ...

File: src/psyclone/tests/psyir/frontend/where_handler_test.py [hunk 6]
```diff
@@ -183,6 +183,9 @@ def test_where_symbol_clash(parser):
     reader = FortranStringReader("module my_test\n"
                                  "contains\n"
                                  "subroutine widx1()\n"
+                                 "  logical, dimension(5,5,5) :: dry\n"
```
Comment (rupertford) [inline src/psyclone/tests/psyir/frontend/where_handler_test.py]: Any particular reason for this, other than more valid code?

File: src/psyclone/tests/psyir/transformations/kernel_transformation_test.py [hunk 7]
```diff
@@ -41,7 +41,7 @@
 import re
 import pytest
 
-from fparser.two.utils import walk_ast
+from fparser.two.utils import walk
```
Comment (rupertford) [inline src/psyclone/tests/psyir/transformations/kernel_transformation_test.py]: 2020 ...

File: src/psyclone/alg_gen.py [hunk 8]
```diff
@@ -221,23 +214,18 @@ def adduse(parse_tree, location, name, only=None, funcnames=None):
...
             break
-        if isinstance(child, (Main_Program, Module, Subroutine_Subprogram,
-                              Function_Subprogram)):
-            parent_prog_statement = child
-
-    if not found:
-        raise GenerationError("alg_gen.py:adduse: The specified location is "
-                              "not in the parse tree.")
-    if not parent_prog_statement:
+        parent = parent.parent
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/alg_gen.py]: Yes, updated.

File: src/psyclone/alg_gen.py [hunk 9]
```diff
@@ -106,13 +106,13 @@ def gen(self):
 
         '''
 
-        from fparser.two.utils import walk_ast
+        from fparser.two.utils import walk
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/alg_gen.py]: Done.

File: .travis.yml [hunk 10]
```diff
@@ -48,7 +48,7 @@ before_install:
   # Specify "fparser_submodule" (without quotes) as an argument to
   # install_optional.sh in order to run the tests with the version of
   # fparser pointed to by the git submodule.
-  - ./bin/install_optional.sh
+  - ./bin/install_optional.sh fparser_submodule
```
[the remark this answers is not in the document]
Comment (arporter) [reply .travis.yml]: Done.

File: examples/nemo/eg3/kernel_utils.py [hunk 11]
```diff
@@ -54,7 +54,7 @@ def valid_kernel(node):
 
     '''
     from psyclone.psyGen import CodeBlock, IfBlock
-    from fparser.two.utils import walk_ast
+    from fparser.two.utils import walk
```
[the remark this answers is not in the document]
Comment (arporter) [reply examples/nemo/eg3/kernel_utils.py]: Done.

File: src/psyclone/alg_gen.py [hunk 12]
```diff
@@ -106,13 +106,13 @@ def gen(self):
...
         # pylint: disable=no-name-in-module
         from fparser.two.Fortran2003 import Call_Stmt, Section_Subscript_List
 
         idx = 0
         # Walk through all statements looking for procedure calls
-        for statement in walk_ast(self._ast.content, [Call_Stmt]):
+        for statement in walk(self._ast.content, Call_Stmt):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/alg_gen.py]: I live for that kind of thing :-)

File: src/psyclone/alg_gen.py [hunk 13]
```diff
@@ -151,45 +151,38 @@ def adduse(parse_tree, location, name, only=None, funcnames=None):
...
 
-    The 'parse_tree' argument is only required as fparser2 currently
-    does not connect a child to a parent. This will be addressed in
-    issue fparser:#102.
-
-    :param parse_tree: The full parse tree of the associated code
-    :type parse_tree: :py:class:`fparser.two.utils.Base`
-    :param location: The current location (node) in the parse tree \
-    provided in the parse_tree argument
+    :param location: The current location (node) in the parse tree to which \
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/alg_gen.py]: Done for all params in this docstring.

File: src/psyclone/alg_gen.py [hunk 14]
```diff
@@ -151,45 +151,38 @@ def adduse(parse_tree, location, name, only=None, funcnames=None):
...
+        to add a USE.
     :type location: :py:class:`fparser.two.utils.Base`
     :param str name: The name of the use statement
     :param bool only: Whether to include the 'only' clause in the use \
-    statement or not. Defaults to None which will result in only being \
-    added if funcnames has content and not being added otherwise.
+        statement or not. Defaults to None which will result in only being \
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/alg_gen.py]: Done.

File: src/psyclone/nemo.py [hunk 15]
```diff
@@ -589,23 +588,14 @@ def match(node):
...
-        array_refs = []
-        if isinstance(rhs, Fortran2003.Part_Ref):
-            # Since walk_ast is slightly clunky we have to manually allow
-            # for the top-level "rhs" node being an array reference
-            array_refs.append(rhs)
-        array_refs += walk_ast(rhs.items, [Fortran2003.Part_Ref])
+        array_refs = walk(rhs, Fortran2003.Part_Ref)
         for ref in array_refs:
-            nested_refs = walk_ast(ref.items, [Fortran2003.Part_Ref])
+            nested_refs = walk(ref.items, Fortran2003.Part_Ref)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/nemo.py]: That's correct. I thought about changing it and decided that it wouldn't be significantly better than what we have.

File: src/psyclone/parse/algorithm.py [hunk 16]
```diff
@@ -41,7 +41,7 @@
 from __future__ import absolute_import
 
 from fparser.two import pattern_tools
-from fparser.two.utils import walk_ast
+from fparser.two.utils import walk
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/parse/algorithm.py]: Done.

File: src/psyclone/parse/kernel.py [hunk 17]
```diff
@@ -46,7 +46,7 @@
 
 from fparser.two.parser import ParserFactory
 from fparser.two import Fortran2003
-from fparser.two.utils import walk_ast
+from fparser.two.utils import walk
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/parse/kernel.py]: Done.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 18]
```diff
@@ -41,7 +41,7 @@
 
 from collections import OrderedDict
 from fparser.two import Fortran2003
-from fparser.two.utils import walk_ast
+from fparser.two.utils import walk
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py]: Done.

File: src/psyclone/tests/alggen_test.py [hunk 19]
```diff
@@ -472,18 +472,19 @@ def get_parse_tree(code, parser):
 
 # function adduse tests These should be moved in #240.
 
-
-def test_adduse_location_none():
+@pytest.mark.parametrize("location", [None, "lilliput"])
+def test_adduse_invalid_location(location):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/alggen_test.py]: Oh yes, well spotted. Fixed now.

File: src/psyclone/tests/alggen_test.py [hunk 20]
```diff
@@ -472,18 +472,19 @@ def get_parse_tree(code, parser):
 
 # function adduse tests These should be moved in #240.
 
-
-def test_adduse_location_none():
+@pytest.mark.parametrize("location", [None, "lilliput"])
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/alggen_test.py]: Yep.

File: src/psyclone/tests/psyir/frontend/where_handler_test.py (hunk 6 above)
Comment (arporter) [reply src/psyclone/tests/psyir/frontend/where_handler_test.py]: Yes, just proofing against the future when the NEMO API will have a symbol table.

**Comment by arporter** [comment]
Assuming Travis is happy then then this is ready for next review.

File: src/psyclone/alg_gen.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/alg_gen.py]: But you didn't do it - boo.

File: src/psyclone/tests/nemo/transformations/loops/loop_test.py [hunk 21]
```diff
@@ -146,14 +146,14 @@ def test_exp_loop_missing_spec(parser):
     fparser2 AST is missing a Specification_Part for the routine.
 
     '''
-    from fparser.two.utils import walk_ast
+    from fparser.two.utils import walk
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/nemo/transformations/loops/loop_test.py]: Thanks

**Pull request merged**